sql/catalyst/src/main/java/org/apache/spark/sql/connector/distributions/ClusteredDistribution.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.connector.distributions;
+
+import org.apache.spark.annotation.Experimental;
+import org.apache.spark.sql.connector.expressions.Expression;
+
+/**
+ * A distribution where tuples that share the same values for clustering expressions are co-located
+ * in the same partition.
+ *
+ * @since 3.2.0
+ */
+@Experimental
+public interface ClusteredDistribution extends Distribution {
```

Comment:
We already have `org.apache.spark.sql.connector.read.partitioning.ClusteredDistribution`, shall we remove that and always use this new one? The new one is better as it's more flexible (using Expression rather than String).

It's better if we can avoid breaking existing APIs, but I can't figure out a good way. The old `ClusteredDistribution` is only used in a mixin trait and marked as `Evolving`. Seems OK to break it.

Reply:
cc @aokolnychyi

Reply:
We have started this discussion in comments to the original PR a bit. Some background [here](https://github.com/apache/spark/pull/29066#issuecomment-735558685) and [here](https://github.com/apache/spark/pull/29066#issuecomment-736073015).

Overall, there are two options: break the API now and use new interfaces instead of the old ones in other places or evolve these two separately for a while and replace the read side once we have a clear plan for bucketed joins.

My first idea was to break the API and migrate to the new interfaces like suggested by @cloud-fan as that seems inevitable at some point. However, there were a couple of good points mentioned by reviewers on the original PR and I am now inclined to evolve these separately for a while until we know what changes will be required to support bucketed tables. I would like to avoid breaking the read side twice if possible.

That said, I don't feel strongly here and each option has its own benefits and drawbacks. I'll be fine with either one.

Reply:
also cc @viirya @HeartSaVioR @rdblue @dongjoon-hyun @sunchao who reviewed the change

Reply:
I'm OK to break it later, just don't leave the old one there forever. Shall we create a blocker ticket for Spark 3.2 to remove the old `ClusteredDistribution`?

Reply:
Sounds ok to me, @cloud-fan.

Reply:
Yes, +1 for filing an JIRA issue to track that and move on for now. Thanks, @cloud-fan and @aokolnychyi .

Reply:
+1 for @dongjoon-hyun's too.

Reply:
I created a blocker SPARK-33807 to track this.

Thanks, everyone!

Reply:
Hi @aokolnychyi , do you have time to clean it up now?

Reply:
Since `Partitioning.satisfy` is still referring to the old interface I'm wondering whether we should defer this to 3.3.0 timeframe. For my storage partitioned join work I found that I need to change the `Partitioning` interface again in a backward incompatible way so that it can work for joins. It may be better to break it once instead of two times. That's my 2 cents.

Reply:
OK, let's retarget it to 3.3

Reply:
Sorry for my late reply. I also assumed we don't have any clarity on how the `Partitioning` interface should look like so I think delaying the update and breaking that API once is better.